File: test/python/bench/test_qpe.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the qpe benchmark."""

from __future__ import annotations

import json
from fractions import Fraction

import pytest

from mqt.core.bench import qpe

from .utils import assert_generates


def test_qpe_accepts_fraction_and_native_phase() -> None:
    """Use exact rational input without a free-form parameter dictionary."""
    options = qpe.Options(
        precision=2,
        phase=Fraction(3, 24),
        method=qpe.Method.ITERATIVE,
    )
    assert options.phase == Fraction(1, 8)

    benchmark = qpe.QPE(options)
    assert benchmark.probability("00") == pytest.approx((2 + 2**0.5) / 8)
    assert benchmark.probability("01") == pytest.approx((2 + 2**0.5) / 8)
    assert json.loads(benchmark.instance_specification_json)["parameters"]["phase"] == {
        "denominator": 8,
        "numerator": 1,
    }

    instance_copy = qpe.QPE.from_instance_specification_json(benchmark.instance_specification_json)
    assert instance_copy.options.phase == Fraction(1, 8)
    assert instance_copy.options.method is qpe.Method.ITERATIVE
    assert instance_copy.case_id == benchmark.case_id

    phase = qpe.Phase(numerator=9, denominator=8)
    native_options = qpe.Options(precision=3, phase=phase)
    assert phase.numerator == 1
    assert phase.denominator == 8
    assert native_options.phase == Fraction(1, 8)
    assert_generates(benchmark.generate())


def test_qpe_rejects_untyped_phase_input() -> None:
    """Reject generic dictionaries at the typed Python boundary."""
    with pytest.raises(TypeError, match=r"fractions\.Fraction or Phase"):
        qpe.Options(
            precision=3,
            phase={"numerator": 1, "denominator": 8},  # ty: ignore[invalid-argument-type]
        )


def test_qpe_normalizes_arbitrary_fraction() -> None:
    """Normalize arbitrary-size fractions before entering the native type."""
    negative = qpe.Options(precision=3, phase=Fraction(-1, 8))
    large = qpe.Options(precision=3, phase=Fraction(2**80 + 1, 8))
    assert negative.phase == Fraction(7, 8)
    assert large.phase == Fraction(1, 8)

    with pytest.raises(ValueError, match="denominator must fit in 64 bits"):
        qpe.Options(precision=3, phase=Fraction(1, 2**80 + 1))

File: test/bench/test_qpe.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/Evaluation.hpp"
#include "bench/QPE.hpp"

#include "support/Diagnostics.hpp"
#include "support/TestSupport.hpp"

#include "gtest/gtest.h"

#include <cmath>
#include <cstddef>
#include <numbers>
#include <string>

namespace {

using mqt::bench::Output;
using mqt::bench::Phase;
using mqt::bench::QPE;
using mqt::bench::QPEMethod;

TEST(Phase, NormalizesTurns) {
  EXPECT_EQ(::mqt::test::value(Phase::create(10, 8)),
            ::mqt::test::value(Phase::create(1, 4)));
  EXPECT_EQ(::mqt::test::value(Phase::create(9, 8)),
            ::mqt::test::value(Phase::create(1, 8)));
  EXPECT_EQ(::mqt::test::value(Phase::create(0, 42)),
            ::mqt::test::value(Phase::create(0, 1)));
  EXPECT_EQ(::mqt::test::errorKind([&] { return Phase::create(1, 0); }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(QPE, UsesDocumentedDefaults) {
  const auto qpe = ::mqt::test::value(QPE::create(
      {.precision = 3, .phase = ::mqt::test::value(Phase::create(3, 8))}));
  EXPECT_EQ(qpe.options().method, QPEMethod::Standard);
  EXPECT_EQ(qpe.output(), (Output{"result", 3}));
}

TEST(QPE, RejectsUnsupportedPrecision) {
  EXPECT_EQ(
      ::mqt::test::errorKind([&] {
        return QPE::create(
            {.precision = 0, .phase = ::mqt::test::value(Phase::create(0, 1))});
      }),
      ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return QPE::create(
                  {.precision = mqt::bench::QPEOptions::MAX_PRECISION + 1,
                   .phase = ::mqt::test::value(Phase::create(0, 1))});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(QPE, RejectsAnUnknownMethod) {
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  constexpr auto invalidMethod = static_cast<QPEMethod>(2);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return QPE::create(
                  {.precision = 2,
                   .phase = ::mqt::test::value(Phase::create(0, 1)),
                   .method = invalidMethod});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(QPE, GivesAnExactDistribution) {
  const auto qpe = ::mqt::test::value(QPE::create(
      {.precision = 3, .phase = ::mqt::test::value(Phase::create(3, 8))}));
  EXPECT_DOUBLE_EQ(::mqt::test::value(qpe.probability("011")), 1.);
  EXPECT_DOUBLE_EQ(::mqt::test::value(qpe.probability("010")), 0.);
  EXPECT_DOUBLE_EQ(::mqt::test::value(qpe.probability("111")), 0.);

  const auto evaluation = ::mqt::test::value(qpe.evaluate({{"011", 100}}));
  EXPECT_DOUBLE_EQ(evaluation.totalVariationDistance, 0.);
  EXPECT_DOUBLE_EQ(evaluation.squaredHellingerFidelity, 1.);
  EXPECT_FALSE(evaluation.successProbability.has_value());
}

TEST(QPE, GivesTheInexactDistribution) {
  const auto qpe = ::mqt::test::value(QPE::create(
      {.precision = 2, .phase = ::mqt::test::value(Phase::create(1, 8))}));
  const auto high = (2. + std::numbers::sqrt2) / 8.;
  const auto low = (2. - std::numbers::sqrt2) / 8.;
  EXPECT_NEAR(::mqt::test::value(qpe.probability("00")), high, 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("01")), high, 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("10")), low, 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("11")), low, 1e-15);
}

TEST(QPE, WrapsTheDistributionAtOneTurn) {
  const auto qpe = ::mqt::test::value(QPE::create(
      {.precision = 2, .phase = ::mqt::test::value(Phase::create(7, 8))}));
  const auto high = (2. + std::numbers::sqrt2) / 8.;
  const auto low = (2. - std::numbers::sqrt2) / 8.;
  EXPECT_NEAR(::mqt::test::value(qpe.probability("00")), high, 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("11")), high, 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("01")), low, 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("10")), low, 1e-15);
}

TEST(QPE, UsesTheNegativeHalfTurnRepresentative) {
  const auto qpe = ::mqt::test::value(QPE::create(
      {.precision = 1, .phase = ::mqt::test::value(Phase::create(7, 8))}));
  EXPECT_NEAR(::mqt::test::value(qpe.probability("0")),
              (2. + std::numbers::sqrt2) / 4., 1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability("1")),
              (2. - std::numbers::sqrt2) / 4., 1e-15);
}

TEST(QPE, SupportsArbitraryWidthOutcomes) {
  constexpr size_t precision = 1025;
  const auto qpe = ::mqt::test::value(QPE::create({
      .precision = precision,
      .phase = ::mqt::test::value(Phase::create(1, 3)),
      .method = QPEMethod::Iterative,
  }));
  auto lower = std::string{};
  lower.reserve(precision);
  for (size_t index = 0; index < precision; ++index) {
    lower.push_back(index % 2 == 0 ? '0' : '1');
  }
  auto upper = lower;
  upper.back() = '1';

  const auto pi = std::numbers::pi;
  EXPECT_NEAR(::mqt::test::value(qpe.probability(lower)), 27. / (16. * pi * pi),
              1e-15);
  EXPECT_NEAR(::mqt::test::value(qpe.probability(upper)), 27. / (4. * pi * pi),
              1e-15);
}

} // namespace
